webapp/app.py

```diff
@@ -1,25 +1,108 @@
-from flask import Flask, jsonify, redirect, render_template, url_for
+import logging
+
+from flask import Flask, Request, jsonify, redirect, render_template, session, url_for
+from flask_login import (
+    LoginManager,
+    UserMixin,
+    login_required,
+    logout_user,
+)
 from werkzeug.wrappers.response import Response
 
+from webapp.config import Config
 from webapp.exceptions import ECSTaskLogStreamDoesNotExistError
-from webapp.utils import get_task_status_and_logs
+from webapp.utils import get_task_status_and_logs, parse_oidc_data
 from webapp.utils.aws import ECSClient
 
+logger = logging.getLogger(__name__)
+
+
+class User(UserMixin):
+    """Class representing users of the app.
+
+    This class stores credentials (MIT ID, name, and email)
+    about logged in users.
+    """
+
+    def __init__(
+        self, id: str, name: str, email: str | None = None  # noqa: A002
+    ) -> None:
+        self.id = id
+        self.name = name
+        self.email = email
+
 
 def create_app() -> Flask:
     app = Flask(__name__)
+    app.secret_key = Config().SECRET_KEY
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.request_loader
+    def load_user_from_request(request: Request) -> User | None:
+        """Callback to login users using data from request headers.
+
+        For this application, login is the process of identifying the user
+        by decoding data provided in the request headers. This method
+        will:
+
+        * Retrieve OpenID Connect (OIDC) access tokens and user data from request headers.
+           * If access tokens or user data are missing from the request headers,
+             user cannot login.
+        * Stores active access tokens and user data in the Flask session.
+        * Create or refresh a User instance with parsed user data from OIDC.
+
+        An expired access token is marked by a mismatch between the OIDC
+        access token retrieved from the request header and the access token
+        currently stored in the Flask session. If there is a mismatch,
+        the method will parse the updated OIDC data in the request header
+        and refresh the User instance with the updated user data.
+        """
+        if (oidc_access_token := request.headers.get("x-amzn-oidc-accesstoken")) is None:
+            logger.error(
+                "Access token ('x-amzn-oidc-accesstoken') is missing "
+                "from the request headers."
+            )
+            return None
+
+        if (oidc_jwt_data := request.headers.get("x-amzn-oidc-data")) is None:
+            logger.error(
+                "User data ('x-amzn-oidc-data') is missing from the request headers."
+            )
+            return None
```

Comment:
I see the reasoning behind the walrus operator here, to check for presence then use the variable later, but this reads a bit awkward to me.  

What about a real boring, but real explicit approach?

```python
oidc_access_token = request.headers.get("x-amzn-oidc-accesstoken")
oidc_jwt_data = request.headers.get("x-amzn-oidc-data")

if not oidc_access_token:
    logger.error(
        "Access token ('x-amzn-oidc-accesstoken') is missing "
        "from the request headers."
    )
    return None

if not oidc_jwt_data:
    logger.error(
        "User data ('x-amzn-oidc-data') is missing from the request headers."
    )
    return None
```

This appeals to me because I can read this top-to-bottom:
1. okay, we're getting important data from headers
2. we're checking they exist and handling if not
3. assume we'll be using those variables down below now that we know they exist

Just a suggestion, as I do think the approach as-is correctly functions.

Reply:
Follow-up: what happens when we return `None`?  Does this instruct `flask-login` that we couldn't parse a user, and therefore access is denied?  That's my assumption, but wanted to confirm.

Also: I do think `logger.error` makes sense, given that we don't expect a request to even make it this far if those headers aren't set.

Reply:
**Follow-up**: Yes, that is correct! A `401 Unauthorized` error is raised and an error page loads.
<img width="1681" alt="image" src="https://github.com/user-attachments/assets/dec3d6fb-d403-43db-b77e-657f154cb25f">

I guess walrus[es] operators are still fresh on the mind, but I agree that the changes you proposed read more cleanly! I'll make the recommended changes.

Reply:
[Walrus operators have been removed](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/60/commits/40227568eb85d1c3cd24e2b5b8be5e89d590fd27#diff-55f12a9f17b4ba61d3e37852150b7d78b8aedbe3629bc4ce8173d4b572297a54R64-R74).